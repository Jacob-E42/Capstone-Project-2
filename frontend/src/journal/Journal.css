/* .JournalEntry {
	margin: 10px;
} */
