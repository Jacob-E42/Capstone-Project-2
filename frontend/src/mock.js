import React, { useCallback, useState } from "react";
import UserContext from "./context_providers/UserContext";
import AlertContext from "./context_providers/AlertContext";
import Request from "./api";
import ApiRequest from "./api";
import ApiContext from "./context_providers/ApiContext";
import JournalContext from "./context_providers/JournalContext";

const demoUser = {
	id: 11,
	firstName: "testfirst",
	lastName: "testlast",
	email: "test@test.net",
	interests: ["baking", "jetskiing"]
};

const demoJournal = {
	id: 42,
	userId: 11,
	title: "Baking adventure",
	entryText:
		"Yesterday I tried baking for the first time like I've always wanted to. It was a complete disaster, but at least it was fun.",
	entryDate: "2023-07-24",
	emotions: null
};

const UserProvider = ({ children }) => {
	return <UserContext.Provider value={{ user: demoUser }}>{children}</UserContext.Provider>;
};

const AnonUserProvider = ({ children }) => {
	const [currentUser, setCurrentUser] = useState(null);

	const login = useCallback(
		async formData => {
			try {
				const request = new Request();
				const token = await request.login(formData);
				if (token) {
					setCurrentUser(demoUser);
					return { success: true };
				} else throw new Error("Log in token in missing");
			} catch (errors) {
				console.error("login failed", errors);
				return { success: false, errors };
			}
		},
		[setCurrentUser]
	);

	const signup = useCallback(
		async formData => {
			try {
				const request = new Request();
				const token = await request.signup(formData);
				console.log(token);
				if (token) setCurrentUser(demoUser);
				else throw new Error("missing sign in token");
				return { success: true };
			} catch (errors) {
				console.error("signup failed", errors);
				return { success: false, errors };
			}
		},
		[setCurrentUser]
	);

	return <UserContext.Provider value={{ currentUser, login, signup }}>{children}</UserContext.Provider>;
};

const AlertProvider = ({ children }) => {
	const [msg, setMsg] = useState("");
	const [color, setColor] = useState("primary");

	return <AlertContext.Provider value={{ msg, setMsg, color, setColor }}>{children}</AlertContext.Provider>;
};

const ApiProvider = ({ children }) => {
	const [request, setRequest] = useState(new ApiRequest());
	request.setToken(
		`eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6MTEsImVtYWlsIjoidGVzdEB0ZXN0Lm5ldCIsImlhdCI6MTY5MjI5MDkxN30.8e12Cw-TMts0FmZALH61ch03kUMlW5QoCi60dlfhyWY`
	);

	return <ApiContext.Provider value={{ api: request }}>{children}</ApiContext.Provider>;
};

const JournalProvider = ({ children }) => {
	return <JournalContext.Provider value={{ journal: demoJournal }}>{children}</JournalContext.Provider>;
};

export { UserProvider, AnonUserProvider, AlertProvider, ApiProvider, JournalProvider };
