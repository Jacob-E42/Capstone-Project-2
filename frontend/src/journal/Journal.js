import React, { useCallback, useContext, useEffect, useState } from "react";
import { Form, FormGroup, Label, Input, Button } from "reactstrap";
import Error from "../common/Error";
import AlertContext from "../context_providers/AlertContext";
import "./Journal.css";

const Journal = ({ date, title, journalType, entryText, setJournal, editJournal }) => {
	console.debug("Journal", date, "Title=", title, "entryText=", entryText);
	const { setMsg, setColor } = useContext(AlertContext);
	let allInfoPresent = date && (title || title === "") && (entryText || entryText === "");
	const [tempJournal, setTempJournal] = useState({
		title: title,
		entryText: entryText,
		journalType: journalType
	});

	useEffect(() => {
		setTempJournal({
			title: title,
			entryText: entryText
		});
	}, [title, entryText]);

	if (!allInfoPresent) {
		setMsg("Required information is missing");
		setColor("danger");
	}

	const handleChange = useCallback(
		async e => {
			console.debug("handleChange");
			e.preventDefault();
			const { name, value } = e.target;
			setTempJournal(tempJournal => ({
				...tempJournal,
				[name]: value
			}));
			console.log(tempJournal);
		},
		[setTempJournal, tempJournal]
	);

	const handleSubmit = useCallback(
		async e => {
			console.debug("handleSubmit");
			e.preventDefault();
			console.log(tempJournal);
			setJournal(tempJournal);

			await editJournal(tempJournal);
		},
		[editJournal, setJournal, tempJournal]
	);

	console.log(title, tempJournal.title, entryText, tempJournal.entryText);
	return (
		<>
			{allInfoPresent && (
				<div className="JournalEntry">
					<header>
						<h2>{date}</h2>
						<Form>
							<FormGroup>
								<Label for="title">Title</Label>
								<Input
									type="text"
									name="title"
									id="title"
									placeholder="title"
									value={tempJournal.title}
									onChange={handleChange}
								/>
							</FormGroup>
						</Form>
						<Form>
							<FormGroup>
								<Label for="journalType">Journal Type</Label>
								<Input
									type="select"
									name="journalType"
									id="journalType"
									value={tempJournal.journalType}
									onChange={handleChange}>
									<option>Daily Journal</option>
									<option>Gratitude Journal</option>
									<option>Reflective Journal</option>
									<option>Stream-of-Consciousness Journal</option>
									<option>Bullet Journal</option>
									<option>Dream Journal</option>
								</Input>
							</FormGroup>
						</Form>
					</header>

					<Form onSubmit={handleSubmit}>
						<FormGroup>
							<Label for="entry">Journal Entry</Label>
							<Input
								type="textarea"
								name="entryText"
								id="entry"
								placeholder="Start your entry here..."
								value={tempJournal.entryText}
								onChange={handleChange}
							/>
						</FormGroup>
						<Button
							type="submit"
							color="success">
							Submit
						</Button>
					</Form>
				</div>
			)}

			{!allInfoPresent && (
				<Error
					msg="Required info is missing"
					color="danger"
				/>
			)}
		</>
	);
};

export default Journal;
