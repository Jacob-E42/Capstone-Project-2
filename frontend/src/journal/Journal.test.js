import React from "react";
import { render, screen, act } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { AlertProvider, ApiProvider, UserProvider } from "../mock";
import Journal from "./Journal";

test("Journal renders without crashing", () => {
	render(
		<MemoryRouter>
			<UserProvider>
				<ApiProvider>
					<AlertProvider>
						<Journal />
					</AlertProvider>
				</ApiProvider>
			</UserProvider>
		</MemoryRouter>
	);
});

test("Journal matches snapshot when rendered correctly", () => {
	const { asFragment } = render(
		<MemoryRouter>
			<UserProvider>
				<ApiProvider>
					<AlertProvider>
						<Journal
							date="2023-07-24"
							title="Baking adventure"
							entryText="Yesterday I tried baking for the first time like I've always wanted to. It was a complete disaster, but at least it was fun."
						/>
					</AlertProvider>
				</ApiProvider>
			</UserProvider>
		</MemoryRouter>
	);

	expect(screen.getByPlaceholderText("title")).toBeInTheDocument();
	expect(screen.getByPlaceholderText("Start your entry here...")).toBeInTheDocument();
	expect(asFragment()).toMatchSnapshot();
});

test("Journal matches snapshot when rendered without props", () => {
	const { asFragment } = render(
		<MemoryRouter>
			<UserProvider>
				<ApiProvider>
					<AlertProvider>
						<Journal />
					</AlertProvider>
				</ApiProvider>
			</UserProvider>
		</MemoryRouter>
	);

	expect(asFragment()).toMatchSnapshot();
});
