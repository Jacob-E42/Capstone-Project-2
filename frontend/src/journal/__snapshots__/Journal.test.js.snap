// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Journal matches snapshot when rendered correctly 1`] = `
<DocumentFragment>
  <div>
    <header>
      <p>
        2023-07-24
      </p>
      <form
        class=""
      >
        <div
          class="mb-3"
        >
          <label
            class="form-label"
            for="title"
          >
            Title
          </label>
          <input
            class="form-control"
            id="title"
            name="title"
            placeholder="title"
            type="text"
            value="Baking adventure"
          />
        </div>
      </form>
    </header>
    <form
      class=""
    >
      <div
        class="mb-3"
      >
        <label
          class="form-label"
          for="entry"
        >
          Journal Entry
        </label>
        <textarea
          class="form-control"
          id="entry"
          name="entry"
          placeholder="Start your entry here..."
        >
          Yesterday I tried baking for the first time like I've always wanted to. It was a complete disaster, but at least it was fun.
        </textarea>
      </div>
      <button
        class="btn btn-success"
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</DocumentFragment>
`;

exports[`Journal matches snapshot when rendered without props 1`] = `
<DocumentFragment>
  <div
    class="alert alert-danger fade"
    role="alert"
  >
    Required info is missing
  </div>
</DocumentFragment>
`;
